### PR TITLE
Upgrade CI to Dart 3.0

### PR DESCRIPTION
**Pull request by isoos**
- Updated the minimum SDK requirement to 3.0, as we want to change some analysis features now that 3.0 is out.
- Also updated end2end version of `dnd` (a follow-up may also update other items in the list).
- Breaking change in clients that may use `PanaProcessResult` or `DartdocResult`.
- Updated dependency test to match the output.

### CODE REVIEW

File: test/goldens/end2end/__timestamp.txt [hunk 1]
```diff
@@ -1 +1 @@
-2022-10-20T07:30:43.697198Z
+2023-05-12T12:55:25.249916Z
```
Comment (jonasfj) [inline test/goldens/end2end/__timestamp.txt:1]: Should we move this to a separate PR?

Comment (isoos) [reply test/goldens/end2end/__timestamp.txt:1]: We want these changes either in this PR or the very next one, because the old version set of the packages and their dependencies doesn't make much sense now with 3.0 SDK.

**Review by sigurdm — APPROVED** [review]
